[clang-tidy] add check to suggest replacement of nested std::min or std::max with initializer lists

sopyb opened the pull request:
closes #25340

Identifies cases where `std::min` or `std::max` is used to find the minimum or maximum value among more than two items through repeated calls. The check replaces these calls with a single call to `std::min` or `std::max` that uses an initializer list. This makes the code slightly more efficient.

sopyb committed:
[clang-tidy] add check to suggest replacement of nested std::min or std::max with initializer lists

sopyb commented:
Since I don't have permission to assign reviewers I will just ping @njames93

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+  const auto *MinCall = Result.Nodes.getNodeAs<CallExpr>("minCall");
+
+  const CallExpr *TopCall = MaxCall ? MaxCall : MinCall;
+  if (!TopCall) {
+    return;
+  }
```
Not needed

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+                                             const Expr **First,
+                                             const Expr **Last,
+                                             std::vector<const Expr *> &Args) {
+  if (!Call) {
+    return;
+  }
```
Not needed, you can take reference instead of pointer and dont worry about checks

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+}
+
+void MinMaxUseInitializerListCheck::findArgs(const CallExpr *Call,
+                                             const Expr **First,
+                                             const Expr **Last,
```
those are out arguments, return them from function as an struct, instead of manipulating arguments.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  const FunctionDecl *Callee = Call->getDirectCallee();
+  if (!Callee) {
+    return;
+  }
```
GENERAL: no need for extra {} for single statement ifs

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.h:
```diff
@@ -0,0 +1,58 @@
+
+namespace clang::tidy::modernize {
+
+/// Transforms the repeated calls to `std::min` and `std::max` into a single
+/// call using initializer lists.
+///
+/// It identifies cases where `std::min` or `std::max` is used to find the
+/// minimum or maximum value among more than two items through repeated calls.
+/// The check replaces these calls with a single call to `std::min` or
+/// `std::max` that uses an initializer list. This makes the code slightly more
+/// efficient.
+/// \n\n
+/// For example:
+///
+/// \code
+///   int a = std::max(std::max(i, j), k);
+/// \endcode
+///
+/// This code is transformed to:
+///
+/// \code
+///   int a = std::max({i, j, k});
+/// \endcode
```
There is no need to keep class description so big, specially that it's check.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.h:
```diff
@@ -0,0 +1,58 @@
+  void registerPPCallbacks(const SourceManager &SM, Preprocessor *PP,
+                           Preprocessor *ModuleExpanderPP) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+
```
consider using TK_IgnoreUnlessSpelledInSource to exclude implicit code.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+MinMaxUseInitializerListCheck::MinMaxUseInitializerListCheck(
+    StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      Inserter(Options.getLocalOrGlobal("IncludeStyle",
+                                        utils::IncludeSorter::IS_LLVM),
+               areDiagsSelfContained()) {}
```
to be honest, because std::min/max is already used, there is no need to check or change includes, as no include need to be added.

sopyb replied:
This is wrong. The overloads required for `std::{min,max}` to accept initializer lists as argument are in `<algorithm>` and it needs to be explicitly imported or there will just be an error that there being no matching function for call.

https://godbolt.org/z/o67qPrG1x

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+        ArgType.getCanonicalType() != ResultType.getCanonicalType();
+
+    if (CastNeeded)
+      ReplacementText += "static_cast<" + ResultType.getAsString() + ">(";
```
may not work for things like `std::vector<int>::size_type`, may put only `size_type`

sopyb replied:
I am unsure if a type cast may even be needed over here since the method declaration for std::{max, min} would require the first 2 parameters being compared would be type casted to the correct type already or it wouldn't compile.

```cpp
template< class T >
const T& max( const T& a, const T& b );

template< class T, class Compare >
const T& max( const T& a, const T& b, Compare comp );

template< class T >
const T& min( const T& a, const T& b );

template< class T, class Compare >
const T& min( const T& a, const T& b, Compare comp );
```

Source: 
 - https://en.cppreference.com/w/cpp/algorithm/min
 - https://en.cppreference.com/w/cpp/algorithm/max

Let me know if I am misunderstanding the intended purpose of the typecast in this case.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+
+void MinMaxUseInitializerListCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *MaxCall = Result.Nodes.getNodeAs<CallExpr>("maxCall");
+  const auto *MinCall = Result.Nodes.getNodeAs<CallExpr>("minCall");
```
code in this method is min/max agnostic. Simply bind both matchers to same name like "call", and then just it here as TopCall

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+                 Lexer::getLocForEndOfToken(TopCall->getEndLoc(), 0,
+                                            Result.Context->getSourceManager(),
+                                            Result.Context->getLangOpts())
+                     .getLocWithOffset(-2)),
```
that -2 may not always work correctly specially where spaces and comments are involved.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,138 @@
+                                            Result.Context->getLangOpts())
+                     .getLocWithOffset(-2)),
+             ReplacementText)
+      << Inserter.createMainFileIncludeInsertion("<algorithm>");
```
as already mention, not needed, we already use std::min, std::max

sopyb replied:
Marking this as resolved please check my other reply [here](https://github.com/llvm/llvm-project/pull/85572#discussion_r1527662792)

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -110,6 +110,12 @@ New checks
+- New :doc:`modernize-min-max-use-initializer-list
+  <clang-tidy/checks/modernize/min-max-use-initializer-list>` check.
+
+  Replaces chained ``std::min`` and ``std::max`` calls that can be written as
+  initializer lists.
```
consider: "... calls with a initializer list where applicable."

PiotrZSL reviewed (changes requested):
Some comments, and check documentation file is not added.

PiotrZSL reviewed (commented):
Tests are also missing.

sopyb committed:
Summary: Refactor min-max-use-initializer-list.cpp and add tests

Details:
- Refactored min-max-use-initializer-list.cpp to handle cases where std::{min,max} is given a compare function argument
- Shortened documentation in MinMaxUseInitializerListCheck.h
- Added tests for min-modernize-min-max-use-initializer-list
- Added documentation for min-modernize-min-max-use-initializer-list
- Updated ReleaseNotes.rst based on mantainer suggestions

sopyb committed:
Squash: Apply Clang format fixes and ensure InnerCall uses std::min or std::max in replacement generation

sopyb commented:
It should be ready for review now. Sorry for requesting while my code was broken. I modified a line and committed it without checking if the code still works.

So there aren't a bunch of "Apply Clang format fixes" I also squished those into a single commit.

sopyb commented:
ping

sopyb commented:
Ping

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,199 @@
+                TopCall->getDirectCallee()->getQualifiedNameAsString() &&
+            (InnerCallNameStr == "std::min" ||
+             InnerCallNameStr == "std::max")) {
+          FindArgsResult innerResult = findArgs(Match, InnerCall);
```
`const FindArgsResult InnerResult`

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.h:
```diff
@@ -0,0 +1,65 @@
+
+namespace clang::tidy::modernize {
+
+/// Replaces chained ``std::min`` and ``std::max`` calls with a initializer list
```
`chained` sound like statement after statement. I think `nested` sounds better, wdyt?
Also: `an initializer list` instead of `a initializer list` (same in release notes and check docs).

Of course, chained `min`/`max` calls could also be detected, but that may be better left for a follow-up pr.

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.h:
```diff
@@ -0,0 +1,65 @@
+  }
+
+private:
+  struct FindArgsResult {
+    const Expr *First;
+    const Expr *Last;
+    const Expr *Compare;
+    std::vector<const Expr *> Args;
+  };
+  utils::IncludeInserter Inserter;
+  FindArgsResult findArgs(const ast_matchers::MatchFinder::MatchResult &Match,
+                          const CallExpr *Call);
+  std::string
+  generateReplacement(const ast_matchers::MatchFinder::MatchResult &Match,
+                      const CallExpr *TopCall, const FindArgsResult Result);
```
These don't need to be members of the check (`Inserter` is just caught in the middle, that one has to stay of course), instead, define `FindArgsResult` and both functions in the cpp file and mark the functions as `static`.

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/modernize/min-max-use-initializer-list.cpp:
```diff
@@ -0,0 +1,105 @@
+// CHECK-MESSAGES: :[[@LINE-1]]:12: warning: do not use nested std::min calls, use std::min({1, 2, 3}, fless_than) instead [modernize-min-max-use-initializer-list]
+// CHECK-FIXES: int min9 = std::min({1, 2, 3}, fless_than);
+
+}
```
Missing newline

5chmidti reviewed (changes requested):
Going in the right direction, but some things still have to be resolved or cleaned up.

Add a few more test (these are the ones I tried locally):
```c++
template <typename T>
struct initializer_list {
  initializer_list()=default;
  initializer_list(T*,int){}
};

template< class T >
const T& max(initializer_list<T> list);
template< class T, class Compare >
const T& max(initializer_list<T> list, Compare comp);
template< class T >
const T& min(initializer_list<T> list);
template< class T, class Compare >
const T& min(initializer_list<T> list, Compare comp);

...

int max2b = std::max(std::max(std::max(1, 2), std::max(3, 4)), std::max(std::max(5, 6), std::max(7, 8)));
// CHECK-MESSAGES: :[[@LINE-1]]:12: warning: do not use nested std::max calls, use std::max({1, 2, 3, 4, 5, 6, 7, 8}) instead [modernize-min-max-use-initializer-list]
// CHECK-FIXES: int max2b = std::max({1, 2, 3, 4, 5, 6, 7, 8});

int max2c = std::max(std::max(1, std::max(2, 3)), 4);
// CHECK-MESSAGES: :[[@LINE-1]]:12: warning: do not use nested std::max calls, use std::max({1, 2, 3, 4}) instead [modernize-min-max-use-initializer-list]
// CHECK-FIXES: int max2c = std::max({1, 2, 3, 4});

int max2d = std::max(std::max({1, 2, 3}), 4);
// CHECK-MESSAGES: :[[@LINE-1]]:12: warning: do not use nested std::max calls, use std::max({1, 2, 3, 4}) instead [modernize-min-max-use-initializer-list]
// CHECK-FIXES: int max2d = std::max({1, 2, 3, 4});

int max2e = std::max(1, max(2, max(3, 4)));
// CHECK-MESSAGES: :[[@LINE-1]]:12: warning: do not use nested std::max calls, use std::max({1, 2, 3, 4}) instead [modernize-min-max-use-initializer-list]
// CHECK-FIXES: int max2e = std::max({1, 2, 3, 4});
```
In the above tests, `max2b` and `max2d` fail, which IMO should work, `max2b` especially.
And please add some tests with type aliases.
My quick test with templates also tells me that they are currently unsupported:
```c++

template <typename T>
void foo(T a, T b, T c){
  auto v = std::max(a, std::max(b, c));
}
```

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,199 @@
+      callExpr(
+          callee(functionDecl(hasName("::std::max"))),
+          anyOf(hasArgument(
+                    0, callExpr(callee(functionDecl(hasName("::std::max"))))),
```
remove duplicates, simply create `auto MaxMatcher = callExpr(callee(functionDecl(hasName("::std::max"))));`
and use it in all those places

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,199 @@
+  const CallExpr *TopCall = Match.Nodes.getNodeAs<CallExpr>("topCall");
+  MinMaxUseInitializerListCheck::FindArgsResult Result =
+      findArgs(Match, TopCall);
+
```
Consider excluding code where TopCall is in macro.

sopyb replied:
@PiotrZSL, could you please provide more details about your review here? I've looked at my code and I haven't used any macros. Could you point out specifically what you're referring to? Thanks for your help.

PiotrZSL replied:
@sopyb It's about test, having simply tests like:
```
#define max3(a, b, c) std::max(a, std::max(b, c))
#define custom_max(a, b) std::max(a, b)
#define value 4545
   std::max(value, std::max(1, 2))
```

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,199 @@
+std::string MinMaxUseInitializerListCheck::generateReplacement(
+    const MatchFinder::MatchResult &Match, const CallExpr *TopCall,
+    const FindArgsResult Result) {
+  std::string ReplacementText =
```
It's should be fine. I worry more about conflicts between checks. I would probably code this using bunch of CreateInsertion, CreateReplacement, ..
As for ReplacementText, SmallString could be used with preallocated like 256B, or simply .reserve

sopyb replied:
> I worry more about conflicts between checks. I would probably code this using bunch of CreateInsertion, CreateReplacement, ..

I remade it to use CreateInsertion, CreateReplacement and CreateRemoval.

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/modernize/min-max-use-initializer-list.cpp:
```diff
@@ -0,0 +1,105 @@
+// CHECK-MESSAGES: :[[@LINE-1]]:12: warning: do not use nested std::min calls, use std::min({4, 5, std::max({2, 3, 1})}) instead [modernize-min-max-use-initializer-list]
+// CHECK-MESSAGES: :[[@LINE-2]]:37: warning: do not use nested std::max calls, use std::max({2, 3, 1}) instead [modernize-min-max-use-initializer-list]
+// CHECK-FIXES: int min3 = std::min({4, 5, std::max({2, 3, 1})});
+
```
add:
```
std::min(std::min(4, 5), std::max(2, utils::max(3, 1)));
```

PiotrZSL reviewed (commented):
Except already raised code-smells, i do not like `MinMaxUseInitializerListCheck::generateReplacement` method.
My main problem with it is that, once things like whitespaces, comments (between arguments) and other stuff will be used, this function may not always produce proper output. As for version 1.0 could be.

sopyb committed:
Addressing code smell issues and consider nested calls with initializer lists as arguments

sopyb committed:
Refactor generateReplacement to return FixItHints to avoid check conflicts

sopyb commented:
Hey there, I've hit a bit of a wall. I'm trying to figure out if a location is inside a comment. The check generates faulty diagnostics, as it is now, if there are comments including '(', '{', '}', and ','. If I could check if the match is inside a comment this issue could be fixed in a few lines of code. I spent all day looking around [doxygen](https://clang.llvm.org/doxygen/) and trying out stuff but nothing seems to work. Here's an example of faulty code:

```cpp
std::min(
  std::min/*some comment containing (*/(
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    // comment containing {
                          ~
    {
      4,
      5 
    // some comment containing }
                               ~
    },
    // some comment containing ,
                               ~
    CmpFunc
    ~~~~~~~ (as intended)
  ),
  ~
  1, 
CmpFunc);
```

so the output is
```cpp
std::min(
  */(
    // comment containing
    {
      4,
      5 
    // some comment containing 
    },
    // some comment containing 
  ,
  1, 
CmpFunc);
```

sopyb commented:
I tried to figure it out today as well, but I couldn't. I just ended up going through the source code figuring out the comment ranges and checking if the characters are contained between those. Let me know if there was a better way to do it and I will go ahead and make more changes.

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/modernize/min-max-use-initializer-list.rst:
```diff
@@ -0,0 +1,26 @@
+
+   int a = std::max(std::max(i, j), k);
+
+`modernize-min-max-use-initializer-list` check will transform the above code to:
```
Add `The` to the start of this sentence, maybe drop the check name.

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,337 @@
+            *Match.SourceManager, Match.Context->getLangOpts())
+            .str();
+
+    if (const auto InnerCall = dyn_cast<CallExpr>(Arg->IgnoreParenImpCasts())) {
```
`const auto *`

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,337 @@
+
+  const auto *TopCall = Match.Nodes.getNodeAs<CallExpr>("topCall");
+
+  FindArgsResult Result = findArgs(TopCall);
```
`const FindArgsResult`

5chmidti reviewed (changes requested):
I've added a comment about the string parsing and comments in `generateReplacement` on how to do this in a better way, which should help you.
Also, if-statements with a single statement branches should not have braces.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,337 @@
+}
+
+static const std::vector<FixItHint>
+generateReplacement(const MatchFinder::MatchResult &Match,
```
this function is over-complicated.
You do not need to do any text parsing, checking characters and so on.
All you need to do is to replace anything that is bettwen arguments with `,` and replace anything bettwen beiging of top call and first argument with `std::min({` and last argument and end with `})`. Comments does not need to be preserved.

sopyb replied:
I've rewritten this as a vector of `FixItHint`s specifically to preserve comments. Right now the only if statement that's there only to preserve comments is on on line 143-146

```cpp
if (utils::lexer::getPreviousToken(
                  InnerResult.Compare->getExprLoc(), *Match.SourceManager,
                  Match.Context->getLangOpts(), false)
                  .getLocation() == Comma->getLocation())
```

to know if there's any comment between the comma and the compare function and create an appropriate removal to preserve the it. I can simplify this and simply remove everything after the last argument in the case there is a compare function (deleting the comment as well), if you'd like me to.

PiotrZSL reviewed (commented):
In short, avoid any string parsing/manipulation.

sopyb committed:
Update documentation

sopyb committed:
consistency changes

sopyb committed:
Remove braces from if statement

sopyb commented:
I am really sorry for the absolute dumpster fire that this PR has been. I had a hard time getting adjusted to the project structure. I genuinely appreciate the feedback and help provided. This was maybe the first time I ever coded in C++ outside small university projects.

I'm pretty sure it is good to go now, but if any issues arise I will come back with the appropriate changes.

sopyb commented:
Please let me know if I should rebase and squash the commits down into a single commit.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,252 @@
+      const FindArgsResult InnerResult = findArgs(InnerCall);
+      const std::vector<FixItHint> InnerReplacements =
+          generateReplacement(Match, InnerCall, InnerResult);
+      if (InnerCall->getDirectCallee()->getQualifiedNameAsString() ==
+              TopCall->getDirectCallee()->getQualifiedNameAsString() &&
+          ((!Result.Compare && !InnerResult.Compare) ||
+           utils::areStatementsIdentical(Result.Compare, InnerResult.Compare,
+                                         *Match.Context))) {
```
invert this if:
```
if (!condition)
   continue;
```

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,252 @@
+  Result.First = nullptr;
+  Result.Last = nullptr;
+  Result.Compare = nullptr;
+
```
`if (Call->getNumArgs() == 0) return ...`
Just in case, or protect on caller level

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,252 @@
+    if (const auto *InitListExpr =
+            dyn_cast<CXXStdInitializerListExpr>(*ArgIterator)) {
+      if (const auto *InitList = dyn_cast<clang::InitListExpr>(
+              InitListExpr->getSubExpr()->IgnoreImplicit())) {
```
what if getSubExpr return nullptr ?

sopyb replied:
> https://github.com/llvm/llvm-project/pull/85572#discussion_r1545536522

I applied the suggestion I was given and I don't think it's possible for it to be nullptr here or I couldn't make it be null in my testing. I can undo it to have the tempExpr.

PiotrZSL reviewed (commented):
Overal looks fine.
Try spliting bigger parts of code int some lambdas or functions, so it could be more readable, consider adding some 1 line comments specially when fixes are created to point out what those insertions/replacement/deletions are.
Check code for missing checks (nullptr, iterators).

In general consider replacing some std::vector with llvm::SmallVector with some preallocated memofry, like for 2 or 32 elements, depend on needs

sopyb committed:
Adding comments and code cleanup

sopyb commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,276 @@
+  const Expr *First;
+  const Expr *Last;
+  const Expr *Compare;
+  SmallVector<const clang::Expr *, 2> Args;
```
Most probably I am only going to have 2 arguments. But in the case the argument is an initializer list the number or args is impossible to know ahead of time.

sopyb committed:
Change ArgText to type StringRef

sopyb committed:
Check if Nested call has no arguments

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,278 @@
+      const auto Comma = utils::lexer::findNextTokenSkippingComments(
+          InnerResult.Last->getEndLoc(), SourceMngr, LanguageOpts);
+
+      // if there are comments between the comma and the comparison
+      if (utils::lexer::getPreviousToken(InnerResult.Compare->getExprLoc(),
+                                         SourceMngr, LanguageOpts, false)
+              .getLocation() != Comma->getLocation()) {
+        // remove the comma and the comparison
+        FixItHints.push_back(
+            FixItHint::CreateRemoval(SourceRange(Comma->getLocation())));
+
+        FixItHints.push_back(
+            FixItHint::CreateRemoval(CharSourceRange::getTokenRange(
+                InnerResult.Compare->getSourceRange())));
+      } else
+        // remove everything after the last argument
+        FixItHints.push_back(
+            FixItHint::CreateRemoval(CharSourceRange::getTokenRange(
+                Comma->getLocation(), InnerResult.Compare->getEndLoc())));
+    }
```
You could simplify this to 
```c++
      // remove the comma and the comparison
      FixItHints.push_back(
          FixItHint::CreateRemoval(SourceRange(Comma->getLocation())));

      FixItHints.push_back(
          FixItHint::CreateRemoval(InnerResult.Compare->getSourceRange()));
```
However, this would leave a whitespace from between the comma and a comparator, but that is not an issue that needs to be resolved by clang-tidy IMO, that is what clang-format is for. 
Thoughts?

sopyb replied:
I was considering it, but I thought it might be a good idea to do a little for code style, but now that I am thinking about it some more would make sense for it to be left for clang-format to fix after I've looked at other linters I have used and how they handle this.

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,278 @@
+  const SmallVector<FixItHint> Replacement =
+      generateReplacement(Match, TopCall, Result);
+
+  if (Replacement.size() <= 2) {
+    return;
+  }
```
It's not immediately clear why `<= 2` is needed, please add a small comment. Also, a single statement if-branch should not have braces.

sopyb replied:
It was to check if the only replacements are the braces then skip the diagnostic could happen if compare functions don't match. Moving those to the check function means I only have to check if the fixes vector is empty.

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,278 @@
+             Match.SourceManager->getFileID(TopCall->getBeginLoc()),
+             "<algorithm>");
+
+  for (const auto &FixIt : Replacement) {
+    Diagnostic << FixIt;
+  }
```
Single statement for-loop -> remove braces

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,278 @@
+
+      // check if there is a comparison argument
+      std::advance(ArgIterator, 1);
+      if (ArgIterator != Call->arguments().end()) {
+        Result.Compare = *ArgIterator;
+      }
```
Single statement if-branch -> remove braces

5chmidti reviewed (changes requested):
Functionality-wise, this looks good to me, I only have some comments regarding cleanup.

Please also add tests for min and max calls that are true-negatives, i.e.: `std::max(1, 2);`, `std::max({1, 2, 3});` and `std::max({1, 2, 3}, less_than);` are probably enough

sopyb commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,278 @@
+
+    if (Result.Compare)
+      FixItHints.push_back(FixItHint::CreateInsertion(
+          Lexer::getLocForEndOfToken(Result.Last->getEndLoc(), 0, SourceMngr,
+                                     LanguageOpts),
```
Weird because on my side I get stuff like `int max2c = std::max({1, 2, 3, }4);` happening where the token would just decide to go before the last argument despite using getEndLoc()

sopyb replied:
I even went ahead and debugged it to make sure Result.Last is the correct Expr and it is. I was unable to figure out why it is happening

5chmidti replied:
You're right. Maybe I switched something up.

sopyb committed:
Addressing issues raised in review

sopyb commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,278 @@
+    // remove the function call
+    FixItHints.push_back(
+        FixItHint::CreateRemoval(CharSourceRange::getTokenRange(
+            InnerCall->getCallee()->getSourceRange())));
```
I genuinely have no idea how I missed this. Thanks for letting me know.

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,255 @@
+    const SmallVector<FixItHint> InnerReplacements =
+        generateReplacement(Match, InnerCall, InnerResult);
+
+    FixItHints.insert(FixItHints.end(),
+                      // ignore { and } insertions for the inner call if it does
+                      // not have an initializer list arg
+                      InnerReplacements.begin(), InnerReplacements.end());
```
The comment no longer applies, and this line can be rewritten as `FixItHints.append(InnerReplacements);` for better readability.

5chmidti commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,255 @@
+          "}");
+  }
+
+  for (const auto &FixIt : Replacement)
+    Diagnostic << FixIt;
```
There is an overload that takes an `ArrayRef<FixItHint>`, so you can write this as `Diagnostic << Replacement;` (nit: `Replacements`)

5chmidti reviewed (approved):
Looks good to me (others are still reviewing), thanks

sopyb committed:
Addressing 5chmidti final suggestions.

HerrCai0907 reviewed (changes requested):
`std::initializer_list` version std::min/std::max will reduce performance in some cases. Please ignore the check for 
1. non-trivial class: https://godbolt.org/z/77ooaGzh6
2. large object: https://godbolt.org/z/14xoz8dnK

sopyb commented:
> `std::initializer_list` version std::min/std::max will reduce performance in some cases. Please ignore the check for
> 
> 1. non-trivial class: https://godbolt.org/z/77ooaGzh6
> 2. large object: https://godbolt.org/z/14xoz8dnK

Thank you for your feedback! I'm grateful for your attention to potential performance impacts related to the use of std::initializer_list with std::min/std::max.

I think the aim of a moderniser check is to enhance code readability and adhere to modern C++ standards, so by default I think modernizing the codebase should be the priority. I'd like to explore the possibility of incorporating performance considerations as an option within the check.

@PiotrZSL, your expertise has been incredibly helpful throughout the process of making my first contribution. I value your perspective on whether we should provide an option within the check to skip addressing code with potential performance implications or have it as the default behavior. This way, users can choose if to prioritize performance or focus solely on code modernization.

I'd love to hear your thoughts on how I should proceed. Thank you once again for your valuable input.

PiotrZSL commented:
@sopyb 
Check is for modernize, not performance, so:
- Add entry in documentation that due to need of copy object into initialization list check may cause performance degradation, add entry that using std::ref, std::cref is recommended in such case:
`b = std::max({std::ref(i), std::ref(j), std::ref(k)});`
- Add option - IgnoreNonTrivialTypes - set by default to true
- Add option - IgnoreTrivialTypesOfSizeAbove - set by default to 32 bytes
Options should be easy to add, check other checks.
If you want quickly deliver version 1.0, then just limit check to built-in types.

As for copies of large types, that's more a thing for new performance check.

sopyb committed:
Add performance options to modernize check

sopyb commented:
Sorry for taking so long to come back with the changes. I have changed my environment last week and didn't have time to properly setup everything to make the required changes until now.

sopyb commented:
Ping

PiotrZSL reviewed (approved):
Except few pointed out nits, looks fine to me.
Would be nice to run it on some bigger project like llvm or something.

sopyb commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,274 @@
+    Result.Compare = *(std::next(Call->arguments().begin(), 2));
+  }
+
+  if (Result.Compare)
+    Result.Args = SmallVector<const Expr *>(llvm::drop_end(Call->arguments()));
+  else
+    Result.Args = SmallVector<const Expr *>(Call->arguments());
```
If I would move it into the if/else. I would have to add another line at the very least `Result.Args = SmallVector<const Expr *>(Call->arguments())` in the other branch so I will not move this into the if/else statement.

PiotrZSL replied:
Why ? You can move line 69 after 65, and 71 after 62, and should be sufficient

sopyb replied:
I am sorry, you're right I forgot I hadnle the initializer list case on it's own and I don't have to worry about the compare potentially being in the arguments

sopyb committed:
Addressing the part of Piotr's review

sopyb commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,274 @@
+                              .concat(ResultType.getAsString(LanguageOpts))
+                              .concat(">(")
+                              .concat(ArgText)
+                              .concat(")");
```
Thanks for the heads up! After taking a look at other checks, it seems like it would be okay to skip assigning `ResultType.getAsString(LanguageOpts)` to a local variable. Storing it locally might  also create some reference lifetime issues.

PiotrZSL replied:
It doesn't mean that other checks are correct. Twine saves only pointer to std::string, it doesn't make a copy. Therefore in this case after call to last concat result of getAsString will be destroyed, if you would add call to .str() at the end of this chain it would be ok, but currently this is just a bug, additionally most probably because return type is small, it fit into stack memory.

PiotrZSL replied:
And this is reported by address sanitizer:
```
==335149==ERROR: AddressSanitizer: stack-use-after-scope on address 0x725ae7653890 at pc 0x56da3506075a bp 0x7ffccd4af960 sp 0x7ffccd4af958
READ of size 8 at 0x725ae7653890 thread T0
    #0 0x56da35060759 in print /fpwork/llvm-project/llvm/lib/Support/Twine.cpp:165:21
    #1 0x56da35060759 in llvm::Twine::printOneChild(llvm::raw_ostream&, llvm::Twine::Child, llvm::Twine::NodeKind) const /fpwork/llvm-project/llvm/lib/Support/Twine.cpp:65:16
    #2 0x56da3505f618 in print /fpwork/llvm-project/llvm/lib/Support/Twine.cpp:165:3
    #3 0x56da3505f618 in toVector /fpwork/llvm-project/llvm/lib/Support/Twine.cpp:34:3
    #4 0x56da3505f618 in llvm::Twine::toStringRef(llvm::SmallVectorImpl<char>&) const /fpwork/llvm-project/llvm/include/llvm/ADT/Twine.h:495:7
    #5 0x56da3505f190 in llvm::Twine::str[abi:cxx11]() const /fpwork/llvm-project/llvm/lib/Support/Twine.cpp:29:10

```

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -80,27 +80,26 @@ static SmallVector<FixItHint>
                      const CallExpr *TopCall, const FindArgsResult &Result,
                      const bool IgnoreNonTrivialTypes,
-                     const unsigned long IgnoreTrivialTypesOfSizeAbove) {
+                     const long IgnoreTrivialTypesOfSizeAbove) {
```
avoid using things like long/int. We got std::uint64_t, std::uint32_t for these things.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.h:
```diff
@@ -47,7 +47,7 @@ class MinMaxUseInitializerListCheck : public ClangTidyCheck {
 private:
   bool IgnoreNonTrivialTypes;
-  unsigned long IgnoreTrivialTypesOfSizeAbove;
+  long IgnoreTrivialTypesOfSizeAbove;
```
why change into signed ? Unsigned were better.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/modernize/min-max-use-initializer-list.rst:
```diff
@@ -0,0 +1,50 @@
+.. option:: IgnoreTrivialTypesOfSizeAbove
+
+   An integer specifying the size (in bytes) above which trivial types are 
+ignored. Default is `32`.
```
add 2 spaces at the begining.

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.h:
```diff
@@ -0,0 +1,56 @@
+
+private:
+  bool IgnoreNonTrivialTypes;
+  long IgnoreTrivialTypesOfSizeAbove;
```
why long ? use std::uint64_t

PiotrZSL reviewed (changes requested):
Fix sanitizer issue.

sopyb committed:
Fix sanitizer issue

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/modernize/min-max-use-initializer-list.rst:
```diff
@@ -0,0 +1,50 @@
+
+.. code-block:: cpp
+
+   int a = std::max(std::max(i, j), k);
```
GENERIC: use 2 spaces instead of 3 for block indent

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/modernize/min-max-use-initializer-list.rst:
```diff
@@ -0,0 +1,50 @@
+
+.. option:: IgnoreTrivialTypesOfSizeAbove
+
+   An integer specifying the size (in bytes) above which trivial types are 
+  ignored. Default is `32`.
```
use same indent, 2 spaces for options

PiotrZSL commented on clang-tools-extra/clang-tidy/modernize/MinMaxUseInitializerListCheck.cpp:
```diff
@@ -0,0 +1,268 @@
+    return FixItHints;
+
+  if (isResultTypeTrivial &&
+      Match.Context->getTypeSizeInChars(ResultType).getQuantity() >
+          IgnoreTrivialTypesOfSizeAbove)
```
`67: warning: comparison of integers of different signs: 'QuantityType' (aka 'long') and 'const std::uint64_t' (aka 'const unsigned long') [-Wsign-compare]`

Just add `static_cast` to `std::uint64_t`

sopyb committed:
Clang-tidy & indenting

PiotrZSL reviewed (approved):
LGTM
